### ISSUE TITLE
Does not work well with lambda functions

### ISSUE THREAD

**Issue by DominikDeak** [sth/callable.hpp#1]
Nice work with this utility, but unfortunately I ran into problems with trying to introspect lambda functions.

I'm using xcode version 6.3.1 (6D1002),  clang C++14.

The following code fails:

``` cpp
auto fn = [](const auto& I) -> int {return I + 100;};

using ReturnType = typename callable_traits<decltype(fn)>::return_type;
```

I get an error,

> Reference to overloaded function could not be resolved; did you mean to call it?

At the following line (72):

``` cpp
struct callable_traits_d : detail::callable_traits_memfn<decltype(&Callable::operator())> {
```

**Comment by maddanio** [comment]
yes, that is because there is an ambiguity. I suppose the whole utility will have problems with polymorphic functions...

**Comment by dimitry-ishenko** [comment]
The title should be renamed to "Does not work with _generic_ lambdas"

**Comment by dimitry-ishenko** [comment]
Come to think of it, I don't think it should. Generic lambdas use templated call operator. So, you are basically saying:

``` c++
    template<typename T> int foo(T) { return 0; }
    using ReturnType = typename callable_traits<decltype(foo)>::return_type;
```

On the other hand, it would be cool if `callable_traits` could take function template.
